test: bsim: Notification EATT test

CarbuneanuAlexandru opened the pull request:
Increasing the test coverage of notification on EATT bearers.
- test implementation

Signed-off-by: Alexandru Carbuneanu <alexandru.carbuneanu@nordicsemi.no>

CarbuneanuAlexandru commented:
@jhedberg can you pls have a look at this PR?

jhedberg reviewed (changes requested):
There seems to be a general issue with reference counting, in particular with `bt_conn` objects. Basically, whenever you copy a pointer to a reference counted object you should be using the appropriate `ref()` function, and whenever you clear a pointer you should use  the appropriate `unref()` function.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt_notif/src/client_test.c:
```diff
@@ -0,0 +1,213 @@
+
+void send_notification(void)
+{
+	uint8_t sample_dat = SAMPLE_DATA;
```
```suggestion
	const uint8_t sample_dat = SAMPLE_DATA;
```

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt_notif/src/client_test.c:
```diff
@@ -0,0 +1,213 @@
+
+	err = bt_enable(NULL);
+	if (err != 0) {
+		FAIL("Bluetooth discover failed (err %d)\n", err);
```
```suggestion
		FAIL("Bluetooth enable failed (err %d)\n", err);
```

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt_notif/src/client_test.c:
```diff
@@ -0,0 +1,213 @@
+
+	local_attr = &g_svc.attrs[1];
+	printk("############# Notification test\n");
+	for (int indx = 0; indx < NUM_NOTIF; indx++) {
```
`indx` is a pretty uncommon abbreviation of `index`. 

Suggest to use `i`, `idx` or `index`.

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt_notif/src/client_test.c:
```diff
@@ -0,0 +1,213 @@
+	}
+
+	printk("############# Disconnect and reconnect\n");
+	for (int indx = 0; indx < CONFIG_BT_EATT_MAX; indx++) {
```
Ditto

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt_notif/src/client_test.c:
```diff
@@ -0,0 +1,213 @@
+
+	WAIT_FOR_FLAG(flag_is_connected);
+
+	err = bt_eatt_connect(g_conn, CONFIG_BT_EATT_MAX);
+	if (err) {
+		FAIL("Sending credit based connection request failed (err %d)\n", err);
+	}
+
+	local_attr = &g_svc.attrs[1];
+	printk("############# Notification test\n");
```
Should we wait for the EATTs to be connected before we start sending notifications?

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt_notif/src/client_test.c:
```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Can you elaborate a bit on this test? 

It seems that it connects multiple EATT and then sends a large number of notifications, but then also starts disconnecting and reconnecting EATT and then attempting to send a notification during discovery. 

What is the purpose of this test?

CarbuneanuAlexandru replied:
Yes, this test is meant to increase the coverage for the notifications over EATT as they should be reliable.
But you can propose more ideas on how to better test the notifications if you think there is not sufficient coverage here.

Thalley replied:
Sorry, I see that my comment was unclear - Can you add information about the test as comments in the file, so that may be more clear in the future what it is testing and why? 

This test is a little less straight forward than most other tests, as it does multiple things to test EATT notifications :)

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt_notif/src/common.c:
```diff
@@ -0,0 +1,64 @@
+{
+	bst_ticker_set_next_tick_absolute(WAIT_TIME);
+	bst_result = In_progress;
+}
+/* Call in init functions*/
```
```suggestion
}

/* Call in init functions*/
```

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt_notif/src/common.c:
```diff
@@ -0,0 +1,64 @@
+/* Call in init functions*/
+void device_sync_init(uint device_nbr)
+{
+	uint peer = CENTRAL_ID;
+
+	if (device_nbr == CENTRAL_ID) {
+		peer = PERIPHERAL_ID;
+	}
+
+	uint dev_nbrs[BACK_CHANNELS] = {peer};
+	uint channel_nbrs[BACK_CHANNELS] = {0};
+	uint *ch = bs_open_back_channel(device_nbr, dev_nbrs, channel_nbrs, BACK_CHANNELS);
```
```suggestion
	const uint peer = CENTRAL_ID;

	if (device_nbr == CENTRAL_ID) {
		peer = PERIPHERAL_ID;
	}

	uint dev_nbrs[BACK_CHANNELS] = { peer };
	uint channel_nbrs[BACK_CHANNELS] = { 0 };
	const uint *ch = bs_open_back_channel(device_nbr, dev_nbrs, channel_nbrs, BACK_CHANNELS);
```

CarbuneanuAlexandru committed:
test: bsim: Notification EATT test

Increasing the test coverage of notification on EATT bearers.
- test implementation

Signed-off-by: Alexandru Carbuneanu <alexandru.carbuneanu@nordicsemi.no>

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt_notif/src/client_test.c:
```diff
@@ -0,0 +1,218 @@
+		struct bt_gatt_discover_params *params)
+{
+	SET_FLAG(flag_discover_complete);
+	printk("Discover complete\n");
+	return BT_GATT_ITER_STOP;
```
```suggestion
	printk("Discover complete\n");
	
	return BT_GATT_ITER_STOP;
```

Thalley commented on tests/bluetooth/bsim_bt/bsim_test_eatt_notif/src/client_test.c:
```diff
@@ -0,0 +1,234 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * EATT notification reliability test:
+ * A central acting as a GATT client scans and connects
+ * to a peripheral acting as a GATT server.
+ * The GATT client will then attempt to connect a number of CONFIG_BT_EATT_MAX bearers
+ * over EATT, send notifications, disconnect all bearers and recconect EATT_BEARERS_TEST
+ * and send start a transaction with a request, then send a lot of notifications
+ * before the response is received.
+ * The test might be exanded by cheking that all the notifications all trasmeted
+ * on EATT channels.
+ */
```
Typos

```suggestion
 * EATT notification reliability test:
 * A central acting as a GATT client scans and connects
 * to a peripheral acting as a GATT server.
 * The GATT client will then attempt to connect a number of CONFIG_BT_EATT_MAX bearers
 * over EATT, send notifications, disconnect all bearers and reconnect EATT_BEARERS_TEST
 * and send start a transaction with a request, then send a lot of notifications
 * before the response is received.
 * The test might be expanded by checking that all the notifications all transmitted
 * on EATT channels.
 */
```

carlescufi commented:
@CarbuneanuAlexandru please rebase

CarbuneanuAlexandru commented:
@jhedberg can you please approve again ?
